### PR TITLE
fix(engine): clear stale token cache on Engine delete-recreate

### LINKED ISSUE #467: BUG: Engine delete/recreate causes permanent "token not authenticated" failure (up to 42 min outage)
## Summary

When an Engine resource is deleted and recreated with the same name, the WASM plugin fails to fetch rules from the cache server with `"token not authenticated"` errors. All traffic returns HTTP 500 (`failurePolicy: fail` with no rules loaded). The condition persists for up to 42 minutes and requires an operator restart to resolve immediately.

## Root cause

When an Engine is deleted, the reconciler cleans up metrics and NetworkPolicies but **does not remove the cached token** from the in-memory `tokenStore`. The token key (`namespace/engineName/rulesetName`) survives deletion.

When a new Engine with the same name is created:
1. `ensureCacheToken()` finds the old entry in `tokenStore`
2. `NeedsRenewal()` returns false (token issued less than 48 min ago)
3. The old token is returned — bound to a now-deleted ServiceAccount
4. `TokenReview` returns "not authenticated" → cache server rejects requests with 403
5. WASM plugin has no rules → HTTP 500 on all traffic
6. Self-heals only when the token hits its 48-min renewal deadline

A secondary issue: `ensureCacheClientServiceAccount()` does not filter out ServiceAccounts with a `DeletionTimestamp` set, allowing a narrow race where a terminating SA could be picked up during GC.

## Affected scenarios

- Manual `kubectl delete` + `kubectl apply`
- OLM upgrade that replaces the Engine resource
- GitOps tools that delete-before-apply (e.g. ArgoCD with `Replace=true`)
- Helm `upgrade --force`

## Reproduction

1. Create an Engine and verify WAF is working (HTTP 200)
2. Delete the Engine and **wait at least 6 minutes** (auth cache must expire)
3. Recreate the Engine with the same name (within 48 min of original token issuance)
4. Traffic returns HTTP 500 immediately

The 6-minute wait is critical — the cache server caches successful `TokenReview` results for 5 minutes. Quick delete/recreate cycles mask the bug because the auth cache still holds a "valid" result from the previous session.

### Cannot reproduce when

- Operator was restarted after deletion (clears in-memory tokenStore)
- Engine recreated within 5 minutes of deletion (auth cache masks the bug)
- More than 48 minutes elapsed since token issuance (renewal generates a fresh token)

## Workaround

```bash
kubectl -n <operator-namespace> rollout restart deploy/coraza-kubernetes-operator


Co-Authored-By: Claude Opus 4.6 <noreply@anthropic.com>

### PR DESCRIPTION
**Pull request by zkraus**
When an Engine is deleted and recreated with the same name, the tokenStore could serve a stale token bound to the now-deleted ServiceAccount. Clear all cached tokens for the Engine in the IsNotFound path and filter out terminating ServiceAccounts when discovering existing ones.

Extract deleteTokensByPrefix helper to share prefix-matching logic with cleanupStaleTokens and unify the matching style.

Add unit test for the full delete-recreate lifecycle verifying token and ServiceAccount renewal.

Fixes #467
Co-Authored-By: Claude Opus 4.6 <noreply@anthropic.com>
Signed-off-by: Zdenek Kraus <zkraus@redhat.com>

## Summary by CodeRabbit

* **Bug Fixes**
  * Removed cached authentication tokens when an Engine is no longer found, preventing stale credentials from being reused.
  * Ignored terminating or previously owned ServiceAccounts when selecting an active account.
  * Ensured recreated Engines receive fresh ServiceAccounts and tokens.
  * Improved outdated-token cleanup while preserving the current token.

* **Tests**
  * Added coverage for Engine recreation, token cleanup, fresh token generation, and stale ServiceAccount handling.

### CODE REVIEW

**Comment by zkraus** [comment]
Rebased onto main. No code change.

**Comment by zkraus** [comment]
Rebased again, because of failed tests. Now with limited GOMAXPROCS hoping for success

File: internal/controller/engine_controller_token.go [hunk 1]
```diff
@@ -111,17 +111,27 @@ func (r *EngineReconciler) ensureCacheClientServiceAccount(ctx context.Context,
...
 	}
-	if len(saList.Items) > 0 {
-		if len(saList.Items) > 1 {
+
+	// Filter out terminating SAs — their tokens will become invalid
+	// once garbage collection completes.
+	var activeSAs []corev1.ServiceAccount
+	for i := range saList.Items {
+		if saList.Items[i].DeletionTimestamp.IsZero() {
+			activeSAs = append(activeSAs, saList.Items[i])
```
[the remark this answers is not in the document]
Comment (zkraus) [reply internal/controller/engine_controller_token.go]: Resolved in a fixup.

File: internal/controller/engine_controller_token.go [hunk 2]
```diff
@@ -212,16 +223,28 @@ func (r *EngineReconciler) pruneExpiredTokens() {
...
+		}
+		return true
+	})
+}
+
 // cleanupStaleTokens removes token entries for RuleSets that an Engine no
 // longer references. When spec.ruleSet.name changes, the old token (keyed by
 // "namespace/engineName/oldRuleSet") would otherwise leak in the sync.Map
 // until it expires.
+// This intentionally duplicates the Range loop from deleteTokensByPrefix
```
Comment (rikatz) [inline internal/controller/engine_controller_token.go]: very nit, but given this repeats with the range above (almost on everything) you could create a wrapper for this function like

```golang
func range(prefix, keep string) func(key, _ any) bool {
	return func(key, _ any) bool {
        if k, ok := key.(string); ok && strings.HasPrefix(k, prefix) && keep != "" && k != keep {
			r.tokenStore.Delete(key)
		}
        return true
    }
}
```

and kinda call it as `r.tokenStore.Range(range(prefix, keep))` that this should work (didn't tested, just thinking on deduplication).

but this is probably too much, I am just being critical on code as requested ;)

File: internal/controller/engine_controller_test.go [hunk 3]
```diff
@@ -2195,6 +2195,269 @@ func TestEngineReconciler_MetricsRecordOnSuccess(t *testing.T) {
...
+func TestEngineReconciler_DeleteRecreateGetsNewToken(t *testing.T) {
+	ctx, cleanup := setupTest(t)
+	defer cleanup()
+
+	const (
+		engineName  = "stale-token-engine"
+		rulesetName = "stale-token-ruleset"
+	)
+
+	createTestGateway(t, ctx, k8sClient, "test-gw", testNamespace)
```
Comment (rikatz) [inline internal/controller/engine_controller_test.go]: as we don't use generateName (https://github.com/networking-incubator/coraza-kubernetes-operator/blob/492955f6eea618d93e9ee7a84152c1b360f48372/internal/controller/engine_controller_test.go#L58) please use a different gateway name here (like delete-recreate-gw) to avoid weird leaks between tests

Comment (rikatz) [reply internal/controller/engine_controller_test.go]: (or even better, stale-token-gw to keep consistency with the other names above)

File: internal/controller/engine_controller_test.go [hunk 4]
```diff
@@ -2195,6 +2195,269 @@ func TestEngineReconciler_MetricsRecordOnSuccess(t *testing.T) {
...
+	// set) — handleNetworkPolicyDeletion removes the finalizer.
+	_, err = reconciler.Reconcile(ctx, req)
+	require.NoError(t, err)
+
+	// Verify the Engine is now fully deleted (finalizer removed → GC'd).
+	var gone wafv1alpha1.Engine
+	require.True(t, apierrors.IsNotFound(k8sClient.Get(ctx, req.NamespacedName, &gone)),
+		"Engine must be gone after finalizer removal")
+
+	// Reconcile the IsNotFound path — cleans up the tokenStore.
```
Comment (rikatz) [inline internal/controller/engine_controller_test.go]: nit:

ask claude to stop adding non-ascii characters:


```suggestion
	// Reconcile the IsNotFound path - cleans up the tokenStore.
```

Comment (rikatz) [reply internal/controller/engine_controller_test.go]: btw this applies everywhere around, for arrows, emdashes, etc

File: internal/controller/engine_controller_test.go [hunk 5]
```diff
@@ -2195,6 +2195,269 @@ func TestEngineReconciler_MetricsRecordOnSuccess(t *testing.T) {
...
+func TestEngineReconciler_StaleOwnerSANotReused(t *testing.T) {
+	ctx, cleanup := setupTest(t)
+	defer cleanup()
+
+	const (
+		engineName  = "stale-owner-engine"
+		rulesetName = "stale-owner-ruleset"
+	)
+
+	createTestGateway(t, ctx, k8sClient, "test-gw", testNamespace)
```
Comment (rikatz) [inline internal/controller/engine_controller_test.go]: same as above, use a different name to avoid colision if possible

File: internal/controller/engine_controller_token.go (hunk 2 above)
Comment (zkraus) [reply internal/controller/engine_controller_token.go]: I am glad that you pointed this out, I thought about it, and I was thinking of something similar, to de-duplicate the code. In the end, I rather wanted to no invent too much code and focus on the tokens. 

Also my thinking was around amount of executions without `keep` and with `keep`, so it wont be slowing down the evaluation of the loop in former case, running additional condition while not necessary. Or am I overthinking optimization due to few ns ?

File: internal/controller/engine_controller_test.go (hunk 3 above)
Comment (zkraus) [reply internal/controller/engine_controller_test.go]: ah, makes sense. Few test reused 'test-gw' claude assumed and presented it is the norm and I did not questioned that. I would go rather with `stale-token-gw`, or maybe it could be matching the test name or some shorter form. Which could lead to use of some `generateName()`. Will change to `delete-recreate-stale-token-gw`

File: internal/controller/engine_controller_test.go (hunk 4 above)
Comment (zkraus) [reply internal/controller/engine_controller_test.go]: great catch. I need to check if the IDE doesn't have ligatures enables, because Jetbrains Pycharm, etc. has that (not sure if default) and it does looks nice, but it should not be in the code itself. Addressed in fixup.

File: internal/controller/engine_controller_test.go (hunk 5 above)
Comment (zkraus) [reply internal/controller/engine_controller_test.go]: ack. Updated both name, with shared suffix.
`delete-recreate-stale-token-gw`
`stale-owner-stale-token-gw`

File: internal/controller/engine_controller_token.go (hunk 2 above)
Comment (zkraus) [reply internal/controller/engine_controller_token.go]: ok, reused your prototype and it should be refactored in a fixup

**Comment by zkraus** [comment]
Pushed non-squashed fixups, maybe for easier review. I worked with claude to perform rather QE PoV review, and some unit tests were added, but the 2 main tests got quite more complex, with additional engine and WASM plugin checks.

I think they check more corner cases, which is nice, but at the same time, those tests starts to be too long, and the previous version, has been testing the original bug quite well.

Also 2 assert helpers were extracted from the additional checks, and may be reusable in the future.

@rikatz I have seen your GW API addition to get rid of `Unstructured`, so here it is with WASM plugin.

File: internal/controller/engine_controller_test.go [hunk 6]
```diff
@@ -2196,8 +2196,7 @@ func TestEngineReconciler_MetricsRecordOnSuccess(t *testing.T) {
 // Engine and recreating it with the same name produces a fresh token bound to
 // a new ServiceAccount, not the stale token from the previous instance.
 func TestEngineReconciler_DeleteRecreateGetsNewToken(t *testing.T) {
-	ctx, cleanup := setupTest(t)
-	defer cleanup()
+	ctx := context.Background()
```
Comment (rikatz) [inline internal/controller/engine_controller_test.go]: nit/followup (no need to change it for now), tests have their own context now (since Go 1.24) that can be used with t.Context() and will respect its own timeout,: https://pkg.go.dev/testing#T.Context

this is more like a nice to have/good cleanup but not really a must for now

File: internal/controller/engine_controller_test.go [hunk 7]
```diff
@@ -492,6 +492,49 @@ func getNestedString(obj map[string]any, key string) (string, bool, error) {
...
+	}, wasmPlugin)
+	require.NoError(t, err)
+
+	spec, found, err := getNestedMap(wasmPlugin.Object, "spec")
+	require.NoError(t, err)
+	require.True(t, found)
+	pluginConfig, found, err := getNestedMap(spec, "pluginConfig")
+	require.NoError(t, err)
+	require.True(t, found)
+	cacheToken, found, err := getNestedString(pluginConfig, "cache_token")
```
Comment (rikatz) [inline internal/controller/engine_controller_test.go]: so, as you are using unstructured, let me help you saving a lot of lines here:

https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#NestedString

tho I don't know why we had this old getNestedXXX functions, and I won't block on this, just wanted to let you know we can clean it later (or leave to use structured API later and make it easier)

File: internal/controller/engine_controller_test.go [hunk 8]
```diff
@@ -492,6 +492,49 @@ func getNestedString(obj map[string]any, key string) (string, bool, error) {
...
+		Name:      name,
+		Namespace: namespace,
+	}, &engine)
+	require.NoError(t, err)
+
+	cond := apimeta.FindStatusCondition(engine.Status.Conditions, conditionType)
+	require.NotNil(t, cond, "Engine must have %s condition", conditionType)
+	assert.Equal(t, expectedStatus, cond.Status,
+		"Engine condition %s expected %s", conditionType, expectedStatus)
+}
```
Comment (rikatz) [inline internal/controller/engine_controller_test.go]: just a simplification, not mandatory.

```suggestion
	assert.True(t, apimeta.IsStatusConditionPresentAndEqual(engine.Status.Conditions, conditionType, expectedStatus))
```

**Review by rikatz — APPROVED** [review]
I have two or 3 simplification comments that can be a followup! thanks!

File: internal/controller/engine_controller_test.go (hunk 8 above)
Comment (zkraus) [reply internal/controller/engine_controller_test.go]: I agree it is a code simplification, but still think it should have and error message to help with the failure analysis. It might be helpful if the apimeta helper could provide a standardized error message right away.

addressed in a fixup

File: internal/controller/engine_controller_test.go (hunk 7 above)
Comment (zkraus) [reply internal/controller/engine_controller_test.go]: addressed in a fixup

File: internal/controller/engine_controller_test.go (hunk 6 above)
Comment (zkraus) [reply internal/controller/engine_controller_test.go]: addressed in a fixup. Background context still used for cleanups.

File: internal/controller/engine_controller_test.go [hunk 9]
```diff
@@ -2192,6 +2213,298 @@ func TestEngineReconciler_MetricsRecordOnSuccess(t *testing.T) {
...
+		ref := metav1.GetControllerOf(&sa)
+		require.NotNil(t, ref, "SA %s must have a controller owner reference", sa.Name)
+		if sa.Name == staleSAName {
+			assert.Equal(t, staleUID, ref.UID,
+				"stale SA must still reference the old Engine UID")
+		} else {
+			assert.Equal(t, engine2.UID, ref.UID,
+				"new SA must be owned by the recreated Engine")
+		}
+	}
```
[the remark this answers is not in the document]
Comment (zkraus) [reply internal/controller/engine_controller_test.go:2488]: Addressed in a fixup.

**Review by rikatz — APPROVED** [review]
thanks!
